File: backend/src/test/kotlin/com/example/backend/service/MovieParserTest.kt
package com.example.backend.service

import com.example.backend.model.Movie
import org.assertj.core.api.Assertions.assertThat
import org.jsoup.Jsoup
import org.junit.jupiter.api.Test
import java.io.File
import java.nio.charset.StandardCharsets

class MovieParserTest {
    
    @Test
    fun `should parse movie data from HTML`() {
        // Given
        val htmlContent = File("src/test/resources/sample-movie-page.html")
            .readText(StandardCharsets.UTF_8)
        val document = Jsoup.parse(htmlContent)
        
        // When
        val movies = MovieParser.parse(document)
        
        // Then
        // Print debug information
        println("Found ${movies.size} movies:")
        movies.forEach { movie ->
            println("Movie: ${movie.title}")
            println("Screenings: ${movie.screenings.size}")
            movie.screenings.forEach { screening ->
                println("  - DateTime: ${screening.dateTime}")
                println("  - Venue: ${screening.venue}")
            }
        }
        
        assertThat(movies).isNotEmpty
            .withFailMessage("No movies were parsed from the HTML")
        
        // Check for specific movie
        val dearEvanHansen = movies.find { it.title == "אוון הנסן היקר" }
        assertThat(dearEvanHansen)
            .withFailMessage("Could not find movie 'אוון הנסן היקר' in parsed movies: ${movies.map { it.title }}")
            .isNotNull
        
        // If found, verify it has screenings
        dearEvanHansen?.let {
            assertThat(it.screenings)
                .withFailMessage("No screenings found for 'אוון הנסן היקר'")
                .isNotEmpty
                
            val firstScreening = it.screenings.first()
            assertThat(firstScreening.dateTime)
                .withFailMessage("Screening datetime is blank")
                .isNotBlank()
            assertThat(firstScreening.venue)
                .withFailMessage("Screening venue is blank")
                .isNotBlank()
        }
        
        // Verify no duplicate movies
        val uniqueTitles = movies.map { it.title }.toSet()
        assertThat(uniqueTitles.size)
            .withFailMessage("Found duplicate movies: ${movies.groupBy { it.title }.filter { it.value.size > 1 }.keys}")
            .isEqualTo(movies.size)
    }

    @Test
    fun `should parse movie 'הבעלים לשעבר' from cinema co il correctly`() {
        // Given
        val htmlContent = File("src/test/resources/cinema_co_il_schedule.html")
            .readText(StandardCharsets.UTF_8)
        val document = Jsoup.parse(htmlContent)
        
        // When
        val movies = MovieParser.parseFromDateHtml(document)

        // Then
        val movie = movies.find { it.title == "הבעלים לשעבר" }
        assertThat(movie).isNotNull
            .withFailMessage("Could not find movie 'הבעלים לשעבר' in parsed movies")
        assertThat(movie?.year).isEqualTo(2023)
        assertThat(movie?.altName).isEqualTo("Ex-Husbands")
        assertThat(movie?.imgUrl).isEqualTo("https://www.cinema.co.il/wp-content/uploads/2024/08/הבעלים-לשעבר.jpg")
        assertThat(movie?.siteUrl).isEqualTo("https://www.cinema.co.il/event/%d7%94%d7%91%d7%a2%d7%9c%d7%99%d7%9d-%d7%9c%d7%a9%d7%a2%d7%91%d7%a8/")
        assertThat(movie?.screenings).anySatisfy { screening ->
            assertThat(screening.dateTime).isEqualTo("2025-01-25 11:30")
            assertThat(screening.venue).isEqualTo("Cinematheque TLV")
        }
    }
} 

File: backend-node/src/__tests__/services/MovieParser.test.ts
import { JSDOM } from 'jsdom';
import fs from 'fs';
import path from 'path';
import { MovieParser } from '../../services/MovieParser';

describe('MovieParser', () => {
  it('should parse movie data from HTML', () => {
    // Given
    const htmlContent = fs.readFileSync(
      path.join(__dirname, '../../__fixtures__/sample-movie-page.html'),
      'utf-8'
    );
    const dom = new JSDOM(htmlContent);
    const document = dom.window.document;

    // When
    const movies = MovieParser.parse(document);

    // Then
    expect(movies).not.toHaveLength(0);

    // Check for specific movie
    const dearEvanHansen = movies.find(movie => movie.title === 'אוון הנסן היקר');
    expect(dearEvanHansen).toBeDefined();

    if (dearEvanHansen) {
      expect(dearEvanHansen.screenings).not.toHaveLength(0);
      
      const firstScreening = dearEvanHansen.screenings[0];
      expect(firstScreening.dateTime).toBeTruthy();
      expect(firstScreening.venue).toBeTruthy();
    }

    // Verify no duplicate movies
    const uniqueTitles = new Set(movies.map(movie => movie.title));
    expect(uniqueTitles.size).toBe(movies.length);
  });

  it('should parse movie with english alt name and metadata correctly', () => {
    // Given
    const htmlContent = fs.readFileSync(
      path.join(__dirname, '../../__fixtures__/cinema_co_il_schedule.html'),
      'utf-8'
    );
    const dom = new JSDOM(htmlContent);
    const document = dom.window.document;

    // When
    const movies = MovieParser.parseFromDateHtml(document);

    // Then
    const movie = movies.find(m => m.title === 'הבעלים לשעבר');
    expect(movie).toBeDefined();
    expect(movie?.year).toBe(2023);
    expect(movie?.altName).toBe('Ex-Husbands');
    expect(movie?.imgUrl).toBe('https://www.cinema.co.il/wp-content/uploads/2024/08/הבעלים-לשעבר.jpg');
    expect(movie?.siteUrl).toBe('https://www.cinema.co.il/event/%d7%94%d7%91%d7%a2%d7%9c%d7%99%d7%9d-%d7%9c%d7%a9%d7%a2%d7%91%d7%a8/');
    
    const screening = movie?.screenings.find(s => 
      s.dateTime === '2025-01-25 11:30' && 
      s.venue === 'Cinematheque TLV'
    );
    expect(screening).toBeDefined();
  });

  it('should parse movie with english alt name from description correctly', () => {
    // Given
    const htmlContent = fs.readFileSync(
      path.join(__dirname, '../../__fixtures__/cinema_co_il_schedule.html'),
      'utf-8'
    );
    const dom = new JSDOM(htmlContent);
    const document = dom.window.document;

    // When
    const movies = MovieParser.parseFromDateHtml(document);

    // Then
    const movie = movies.find(m => m.title === 'הסיפור של סולימן');
    expect(movie).toBeDefined();
    expect(movie?.altName).toBe('The Story of Souleymane');
  });

  it('should parse at least 5 movies from another examples fixture', () => {
    // Given
    const htmlContent = fs.readFileSync(
      path.join(__dirname, '../../__fixtures__/cinema_co_another_examples.html'),
      'utf-8'
    );
    const dom = new JSDOM(htmlContent);
    const document = dom.window.document;

    // When
    const movies = MovieParser.parseFromDateHtml(document);

    // Then
    expect(movies.length).toBeGreaterThanOrEqual(5);
    
    // Verify no duplicate movies
    const uniqueTitles = new Set(movies.map(movie => movie.title));
    expect(uniqueTitles.size).toBe(movies.length);
    
    // Log the parsed movies for debugging
    console.log(`Parsed ${movies.length} movies:`, movies.map(m => m.title));
  });

  it('should parse specific movie with all metadata', () => {
    // Given
    const htmlContent = fs.readFileSync(
      path.join(__dirname, '../../__fixtures__/cinema_co_another_examples.html'),
      'utf-8'
    );
    const dom = new JSDOM(htmlContent);
    const document = dom.window.document;

    // When
    const movies = MovieParser.parseFromDateHtml(document);

    // Then - Find the specific movie
    const movie = movies.find(m => m.title === 'סוף שבוע בגליל | תרגום לצרפתית');
    
    expect(movie).toBeDefined();
    expect(movie?.title).toBe('סוף שבוע בגליל | תרגום לצרפתית');
    
    // Log all data we captured for this movie
    console.log('Movie data:', JSON.stringify(movie, null, 2));
    
    // Check for screenings
    expect(movie?.screenings).toBeDefined();
    expect(movie?.screenings.length).toBeGreaterThan(0);
  });
}); 